src/libxtp/eigencuda.cc

```diff
@@ -0,0 +1,156 @@
+ * two matrices.
+ */
+void CudaPipeline::gemm(const CudaMatrix &A, const CudaMatrix &B,
+                        CudaMatrix &C) const {
+
+  // Scalar constanst for calling blas
+  double alpha = 1.;
+  double beta = 0.;
+  const double *palpha = &alpha;
+  const double *pbeta = &beta;
+
+  if ((A.cols() != B.rows())) {
+    throw std::runtime_error("Shape mismatch in Cublas gemm");
+  }
+
+  cublasDgemm(_handle, CUBLAS_OP_N, CUBLAS_OP_N, A.rows(), B.cols(), A.cols(),
+              palpha, A.pointer(), A.rows(), B.pointer(), B.rows(), pbeta,
+              C.pointer(), C.rows());
+}
+
+/*
+ * \brief Perform a Tensor3D matrix multiplication
+ */
+void CudaPipeline::right_matrix_tensor_mult(
+    std::vector<Eigen::MatrixXd> &tensor, const Eigen::MatrixXd &B) const {
+  // First submatrix from the tensor
+  const Eigen::MatrixXd &submatrix = tensor[0];
+
+  // sizes of the matrices to allocated in the device
+  size_t size_A = submatrix.size() * sizeof(double);
+  size_t size_B = B.size() * sizeof(double);
+  size_t size_C = submatrix.rows() * B.cols() * sizeof(double);
+  throw_if_not_enough_memory_in_gpu(size_A + size_B + size_C);
+
+  // Matrix in the Cuda device
+
+  CudaMatrix matrixA(submatrix.rows(), submatrix.cols());
+  CudaMatrix matrixB{B, _stream};
+  CudaMatrix matrixC(submatrix.rows(), B.cols());
+
+  // Call tensor matrix multiplication
+  for (auto i = 0; i < static_cast<int>(tensor.size()); i++) {
+    // Copy tensor component to the device
+    checkCuda(cudaMemcpyAsync(matrixA.pointer(), tensor[i].data(), size_C,
+                              cudaMemcpyHostToDevice, _stream));
+
+    // matrix multiplication
+    gemm(matrixA, matrixB, matrixC);
+
+    // Copy the result to the host
+    double *hout = tensor[i].data();
+    checkCuda(cudaMemcpyAsync(hout, matrixC.pointer(), size_C,
+                              cudaMemcpyDeviceToHost, _stream));
+  }
+}
+
+/*
+ * \brief performs a matrix_1 * matrix2 * matrix_2 multiplication
+ */
+Eigen::MatrixXd CudaPipeline::triple_matrix_mult(const CudaMatrix &A,
+                                                 const Eigen::MatrixXd &matrix,
+                                                 const CudaMatrix &C) const {
+
+  // sizes of the matrices to allocated in the device
+  size_t size_B = matrix.size() * sizeof(double);
+  std::size_t size_W = A.rows() * matrix.cols() * sizeof(double);
+  std::size_t size_Z = A.rows() * C.cols() * sizeof(double);
+  throw_if_not_enough_memory_in_gpu(size_B + size_W + size_Z);
+
+  // Intermediate Matrices
+  CudaMatrix B{matrix, _stream};
+  CudaMatrix W{A.rows(), matrix.cols()};
+  CudaMatrix Z{A.rows(), C.cols()};
+
+  Eigen::MatrixXd result = Eigen::MatrixXd::Zero(A.rows(), C.cols());
+
+  // Call the first tensor matrix multiplication
+  gemm(A, B, W);
+
+  // Call the second tensor matrix multiplication
+  gemm(W, C, Z);
+
+  // Copy the result Array back to the device
+  checkCuda(cudaMemcpyAsync(result.data(), Z.pointer(), size_Z,
```

Comment:
here again if you create the assignment operator this call will go away.

Reply:
done